File: JavaScript/permutation-palindrome.js
// --- Permutation Palindrome -------------------------------------
// Write an efficient function that checks whether any permutation of an input string is a palindrome.
// You can assume the input string only contains lowercase letters.

function buildLetterCount(string) {
  var letterCounter = {};
  for (var i=0; i < string.length; i++) {
    letterCounter[string[i]] == undefined ? letterCounter[string[i]] = 1 : letterCounter[string[i]] += 1;
  }
  return letterCounter;
}

function isPermutationPalindrome(string){
  if (string.length <= 1) { return true; }

  var letterCounts = Object.values(buildLetterCount(string));
  if (letterCounts.every(function(elem){ return elem % 2 == 0; })) {
    return true;
  } else if (string.length % 2 == 0) {
    return false;
  } else {
    var oddCount = 0;
    for (var j=0; j < letterCounts.length; j++) {
      if (letterCounts[j] % 2 != 0) { oddCount += 1; }
    }
    return oddCount <= 1;
  }
}

// --- Tests -----------------------------------------------------

function runTests(){
  function assert(test, args) {
    if (!test) {
      console.log(args.testNumber + ". " + args.message + ": false");
    } else {
      console.log(args.testNumber + ". " + args.message + ": true");
    };
  };

  assert(
    isPermutationPalindrome('c') === true,
    { message: "'c' is a palindrome", testNumber: 0 }
  );

  assert(
    isPermutationPalindrome('cici') === true,
    { message: "'cici' has a palindrome permutation", testNumber: 1 }
  );

  assert(
    isPermutationPalindrome('civi') === false,
    { message: "'civi' does not have a palindrome permutation", testNumber: 2 }
  );

  assert(
    isPermutationPalindrome('civic') === true,
    { message: "'civic' is a palindrome", testNumber: 3 }
  );

  assert(
    isPermutationPalindrome('ivicc') === true,
    { message: "'ivicc' has a palindrome permutation", testNumber: 4 }
  );

  assert(
    isPermutationPalindrome('civil') === false,
    { message: "'civil' does not have a palindrome permutation", testNumber: 5 }
  );

  assert(
    isPermutationPalindrome('livci') === false,
    { message: "'livci' does not have a palindrome permutation", testNumber: 6 }
  );

  assert(
    isPermutationPalindrome('madam im adam') === true,
    { message: "'madam im adam' is a palindrome", testNumber: 7 }
  );
};

runTests();

File: Python/permutation-palindrome.py
# --- Permutation Palindrome -------------------------------------
# Write an efficient function that checks whether any permutation of an input string is a palindrome.
# You can assume the input string only contains lowercase letters.

def buildLetterCount(string):
    countDictionary = {}
    letterArray = list(string)

    for letter in letterArray:
        if letter in countDictionary:
            countDictionary[letter] += 1
        else:
            countDictionary[letter] = 1

    return countDictionary

def isPermutationPalindrome(string):
    stringLength = len(string)
    if stringLength <= 1:
        return True

    letterCounts = list(buildLetterCount(string).values())
    if all(elem % 2 == 0 for elem in letterCounts):
        return True
    elif len(string) % 2 == 0:
        return False
    else:
        # if word length is not even, permit a single odd count
        oddCount = 0
        for count in letterCounts:
            if count % 2 != 0:
                oddCount += 1

        return False if oddCount > 1 else True



# --- Tests -----------------------------------------------------
import unittest

class TestIsPermutationPalindrome(unittest.TestCase):
    def test_returnsTrueForC(self):
        self.assertEqual(isPermutationPalindrome('c'), True)

    def test_returnsTrueForCici(self):
        self.assertEqual(isPermutationPalindrome('cici'), True)

    def test_returnsFalseForCivi(self):
        self.assertEqual(isPermutationPalindrome('civi'), False)

    def test_returnsTrueForCivic(self):
        self.assertEqual(isPermutationPalindrome('civic'), True)

    def test_returnsTrueForIvicc(self):
        self.assertEqual(isPermutationPalindrome('ivicc'), True)

    def test_returnsFalseForCivil(self):
        self.assertEqual(isPermutationPalindrome('civil'), False)

    def test_returnsFalseForLivci(self):
        self.assertEqual(isPermutationPalindrome('livci'), False)

if __name__ == '__main__':
    unittest.main()
